doc/packages/build-support.md

```diff
@@ -0,0 +1,97 @@
+# Build Support {#sec-build-support}
+
+## `pkgs.substitute` {#pkgs-substitute}
+
+`pkgs.substitute` is a wrapper around [the `substitute` function](#fun-substitute) provided by the
+stdenv. It replaces strings in `src` as specified by the `substitutions` argument.
+
+In a build script, the line:
+
+```bash
+substitute $infile $outfile --replace-fail @foo@ ${foopkg}/bin/foo
+```
+
+is equivalent to:
+
+```nix
+{ substitute, foopkg }:
+substitute {
+  src = ./sourcefile.txt;
+  substitutions = [
+    "--replace"
+    "@foo@"
+    "${foopkg}/bin/foo"
+  ];
+}
+```
+
+## `pkgs.substituteAll` {#pkgs-substituteall}
+
+`pkgs.substituteAll` substitutes all instances of `@varName@` (`@`s included)
+in file file `src` with the value of the corresponding environment variable. As
+this uses the `substituteAll` function, its limitations regarding variable names
+that will or will not be replaced also apply here, and familiarity with [the
+function's documentation](#fun-substitute) is recommended.
+
+For example, if `say-goodbye.sh` contained the following:
+
+```
+#! @bash@/bin/bash
+
+echo @unchanged@
+@hello@/bin/hello --greeting @greeting@
+```
+
+the following derivation would make substitutions to `@bash@`, `@hello@`, and `@greeting@`:
+
+```nix
+{
+  substituteAll,
+  bash,
+  hello,
+}:
+substituteAll {
+  src = ./say-goodbye.sh;
+  env = {
+    inherit bash hello;
+    greeting = "goodbye";
+  };
+}
+```
+
+such that `$out` would result in something like the following:
+
+```
+#! /nix/store/s30jrpgav677fpc9yvkqsib70xfmx7xi-bash-5.2p26/bin/bash
+
+echo @unchanged@
+/nix/store/566f5isbvw014h7knmzmxa5l6hshx43k-hello-2.12.1/bin/hello --greeting goodbye
+```
+
+## `pkgs.substituteAllFiles` {#pkgs-substituteallfiles}
+
+`pkgs.substituteAllFiles` replaces `@varName@` with the value of the environment
+variable `varName`. It expects `src` to be a directory and requires a `files`
+argument that specifies which files will be subject to replacements; only these
+files will be placed in `$out`.
+
+As it also uses the `substituteAll` function, it is subject to the same limitations on environment
+variables as discussed in [pkgs.substituteAll](#pkgs-substituteall).
```

Comment:
As mentioned by @hsjobeki, don't expect prior knowledge or any particular order of presentation (it's still planned to break the manual into pages, things like "prior" or "above" won't make sense there). We can assume most readers will end up right here through some link. Just link to related information directly, using a self-explanatory link text.

Reply:
The highlighted portion is indeed a link and the words `prior` and `above` don't appear in this file at all -- was this just a general admonition? Sorry if I'm missing your suggestion here.